### DISCUSSION TITLE
$fetch - nuxt instance unavailable

### DISCUSSION THREAD

**Discussion by mihail727** [nuxt/nuxt#17811] [Questions]
I'm trying to write a handler that, when receiving status 401 (access token invalid) from the server, updates the tokens and repeats the request
`@/composables/useRequest.ts`
```ts
export const useRequest = async <T>(API: string, method: 'GET' | 'POST' = 'GET', body = null) => {
	try {
		const data = await request(API, method, body);

		if (process.client) {
			return data as unknown as T;
		} else {
			return JSON.parse(data) as T;
		}
	} catch (e) {
		throw new Error(e.message);
	}
};

async function request(API: string, method: 'GET' | 'POST' = 'GET', body = null) {
	return await $fetch<string>(API, {
		method,
		headers: useRequestHeaders(['cookie']),
		body,
		async onResponseError({ response }) {
			if (response.status === 401) {
				console.log('auth');
				await useRequest('/api/auth/refresh');
				return await request(API, method, body);
			} else {
				throw new Error(response._data.message);
			}
		},
	});
}
```
`@/server/middleware/access-middleware.ts`
```ts
import type { IncomingMessage, ServerResponse } from 'http';
import { useCookies } from 'h3';
import { TokenService } from '../services';

const INCLUDED_URLS = ['/api/get-items', '/api/get-users'];

export default async (req: IncomingMessage, res: ServerResponse) => {
	if (INCLUDED_URLS.includes(req.url)) {
		const userData = useCookies(req) as unknown as IUserClient;

		if (!userData.accessToken || !userData.refreshToken) {
			console.log('[Access Middleware]\n Некорректные значения cookies');
			res.writeHead(302, {
				location: '/auth/login',
			});
			res.end();
		}

		if (!(await TokenService.verifyAccessToken(userData.accessToken))) {
			res.statusCode = 401;
			res.end();
		}
	} else return;
};
```
---
How it works:
1) The first request is sent, with invalid tokens
2) the middleware aborts the request and sends a 401 status code
3) the client receives an error and sends a request to refresh the token
4) repeats the request
---
What's happening:
1) an error is shown in the server console: nuxt instance unavailable
2) in the browser console: only the error text is displayed, the data on the page is not displayed
3) tokens are updated both in the database and in cookies
---
Please tell me what I'm doing wrong and how can I rewrite it?

**Comment by Diizzayy** [comment]
@mihail727 I've looked at your code and thus far it looks alright. The composable is written properly as is your middleware. So that leaves your `TokenService` and your SFC.

I suspect this issue is caused elsewhere in your code. Perhaps we'd be able to offer more assistance with additional information as well as seeing more code.

**Reply by mihail727** [reply]
> @mihail727 I've looked at your code and thus far it looks alright. The composable is written properly as is your middleware. So that leaves your `TokenService` and your SFC.
> 
> I suspect this issue is caused elsewhere in your code. Perhaps we'd be able to offer more assistance with additional information as well as seeing more code.

`@/pages/index.vue`
```ts
<script setup lang="ts">
	const users = await useControllers().userController.getUsers();

	definePageMeta({
		middleware: ['is-admin'],
	});
</script>

<template>
	<div>{{ users }}</div>
</template>
```
`@/controllers/userController.ts`
```ts
async getUsers() {
	try {
		const users = await useRequest<IUserClient[]>('/api/get-users');
		return users;
	} catch (e) {
		console.error(`[User Controller][Get Users] ${e.message}`);
		return [];
	}
}
```
`@/server/api/refresh.ts`
```ts
import { CookiesController } from './../../controllers/cookies-controller';
import { UserService } from './../../services/user-service';
import type { IncomingMessage, ServerResponse } from 'http';
import { useCookie } from 'h3';

export default async (req: IncomingMessage, res: ServerResponse) => {
	try {
		const refreshToken = useCookie(req, 'refreshToken');
		console.log(refreshToken);
		const response = await UserService.refresh(refreshToken);
		CookiesController.setResponseCookies(res, response);
		res.end(JSON.stringify(response));
	} catch (e) {
		console.log(`[Refresh] \n${e.message}`);
		res.statusCode = 400;
		res.end();
	}
};
```

**Reply by mihail727** [reply]
The code is not the most attractive and I see a lot of erroneous logic in it and I plan to rewrite it in the future, but the essence remains the same, why are the tokens updated, but the data is not displayed? and an error is thrown

**Reply by mihail727** [reply]
I restarted the project and now it seems that this function does not run on the client at all...
i can't run the debugger
![](https://i.imgur.com/loAEyvn.png)

**Reply by Diizzayy** [reply]
Ahh interesting. I'm sure that your composable isn't the problem.

@mihail727 Also, if you have discord.. feel free to add me Diizzayy#1964
